src/server/auth/server/api_server.go

```diff
@@ -1416,22 +1368,147 @@ func (a *apiServer) RestoreAuthToken(ctx context.Context, req *auth.RestoreAuthT
 		}
 	}
 
-	// Check whether the token hash already exists - we don't want to replace an existing token
-	if _, err := col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
-		tokens := a.tokens.ReadWrite(stm)
-		var existing auth.TokenInfo
-		err := tokens.Get(req.Token.HashedToken, &existing)
-		if err == nil {
+	if err := func() error {
+		// Check whether the token hash already exists - we don't want to replace an existing token
+		if _, err := a.lookupAuthTokenInfo(ctx, req.Token.HashedToken); err == nil {
 			return errors.New("cannot overwrite existing token with same hash")
-		} else if err != nil && !col.IsErrNotFound(err) {
+		} else if !col.IsErrNotFound(err) {
 			return err
 		}
-
-		return tokens.PutTTL(req.Token.HashedToken,
-			req.Token.TokenInfo,
-			ttl)
-	}); err != nil {
+		// insert token
+		if ttl > 0 {
+			return a.upsertAuthToken(ctx, req.Token.HashedToken, req.Token.TokenInfo.Subject, ttl)
+		} else {
+			return a.upsertAuthTokenNoTTL(ctx, req.Token.HashedToken, req.Token.TokenInfo.Subject)
+		}
+	}(); err != nil {
 		return nil, errors.Wrapf(err, "error restoring auth token")
 	}
+
 	return &auth.RestoreAuthTokenResponse{}, nil
 }
+
+// returns -1 if token is not found, and the TTL otherwise
+func (a *apiServer) lookupAuthTokenTTLSeconds(ctx context.Context, tokenHash string) (int64, error) {
+	var ttl float64
```

Comment:
What do you think about changing the response struct to return the absolute expiration times instead of a TTL? The reason we expose a TTL in the first place is because that's how etcd did it, but if postgres knows the exact timestamp we could just send that to the user and it seems more useful.

Reply:
good idea